node_modules/gh-orgunit/lib/internal/import.js

```diff
@@ -0,0 +1,594 @@
+    if (validator.hasErrors()) {
+        return callback(validator.getErrors());
+    }
+
+    // Ensure the organisational unit exists
+    OrgUnitAPI.getOrgUnit(ctx, id, false, function(err, orgUnit) {
+        if (err) {
+            return callback(err);
+        }
+
+        // We only support importing data on `timetable` applications
+        if (orgUnit.App.type !== AppsConstants.types.TIMETABLE) {
+            return callback({'code': 400, 'msg': 'This type of application does not allow you to import data'});
+        }
+
+        // We only support imports on `course`, `subject` or `part` nodes
+        if (orgUnit.type !== 'course' && orgUnit.type !== 'subject' && orgUnit.type !== 'part') {
+            return callback({'code': 400, 'msg': 'We only support importing data on courses, subjects or parts'});
+        }
+
+        // Ensure the current user can import data into the organisational unit
+        OrgUnitAuthz.canUpdateOrgUnit(ctx, orgUnit, function(err, canImport) {
+            if (err) {
+                return callback(err);
+            } else if (!canImport) {
+                log().warn({'id': orgUnit.id}, 'Unauthorized attempt at importing an organisational unit');
+                return callback({'code': 401, 'msg': 'You are not allowed to import the organisational unit'});
+            }
+
+            getParentIfNecessary(orgUnit, function(err, parent) {
+                if (err) {
+                    return callback(err);
+                }
+
+                // Prior to merging, verify the external ids are unique WITHIN the import tree. If this
+                // weren't the case, this would get lost in the merge process which would be really painful
+                // to debug. We need to have validated that `data` is an object first though
+                _validateImportDataExternalId(validator, data);
+                if (validator.hasErrors()) {
+                    return callback(validator.getErrors());
+                }
+
+                OrgUnitDAO.exportOrgUnit(orgUnit, 'json', function(err, exportedOrgUnit) {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    // Overlay the import tree over the data that's already present in the database
+                    var mergedData = mergeTrees(exportedOrgUnit, data, deleteMissing);
+
+                    // Ensure the imported data doesn't create validation errors
+                    _validateImport(validator, mergedData.tree, parent);
+                    if (validator.hasErrors()) {
+                        return callback(validator.getErrors());
+                    }
+
+                    // Import the merged data and remove the data that needs to be removed (if any)
+                    return OrgUnitDAO.importTree(mergedData, callback);
+                });
+            });
+        });
+    });
+};
+
+var getParentIfNecessary = function(orgUnit, callback) {
```

Comment:
Needs docs. Also prefix w/ `_`.

FWIW, when OAuth lands this can also be `_.ghIf(orgUnit.parentId, _getParent, orgUnit, function(err, parent) { ...`

:trollface: